src/main/js/utils/generate.js

```diff
@@ -6,57 +6,109 @@ import {
-      const scheduleMaker = new ScheduleMaker(filteredCourseList, parseInt(values.gap, 10));
-      try {
-        schedules = scheduleMaker.makeSchedules();
-      } catch (e) {
-        schedules = e;
-      }
-    }
-    const queryObject = {
-      ...values,
-      c: courseList.map((course) => {
-        const name = `${course[0].subject}${course[0].courseNumber}`;
-        return JSON.stringify({
-          name,
-          ...(values.courses ? values.courses[`${name}${course.id}`] : {}),
-        });
-      }),
-    };
-    delete queryObject.courses;
-    (loadQuery ? Promise.resolve() : fetch(`${shortenUrl}?${stringify({ data: JSON.stringify(queryObject), prefix: 'generator' })}`))
-      .then(response => (loadQuery ? loadQuery.substring(2) : response.json()))
-      .then((result) => {
-        const querystring = `q=${result}`;
-        dispatch(startRedirect({
-          pathname: '/generator',
-          hash: schedules.length > 0 ? '#schedules' : '#timetable',
-          search: result ? querystring : '',
-        }));
-      })
-      .catch(() => {
-        dispatch(startRedirect({
-          pathname: '/generator',
-          hash: schedules.length > 0 ? '#schedules' : '#timetable',
-          search: '',
+    let filteredCourseList = filteredResults.map(result => result.filtered);
+
+    const gpasLoading = values.sortByGPA === 'yes' ? addGPA(filteredCourseList) : Promise.reject();
+
+    gpasLoading
+      .then((res) => {
+        dispatch(addGrades(res));
+        filteredCourseList = filteredCourseList.map(list => list.sort((course1, course2) => {
+          const name = `${course1.subject}${course1.courseNumber}`;
+          const instructor1 = getInstructorLastName(course1.instructor); // Backend only holds instructors' last name
+          const instructor2 = getInstructorLastName(course2.instructor);
+
+          /** Can't sort if the backend doesn't hold data for the entire course */
+          if (!res[name]) return 0;
+
+          /** Assume when instructor === 'Staff', then the course has a gpa of 0.00 */
+          if (instructor1 === 'Staff' && instructor2 === 'Staff') return 0;
+          else if (instructor1 === 'Staff') return 1; // eslint-disable-line
+          else if (instructor2 === 'Staff') return -1;
+
+          /** Use mapping to determine sort order */
+          if (!res[name][instructor1]) return 1;
+          else if (!res[name][instructor2]) return -1; // eslint-disable-line
+          else {
+            const sum1 = res[name][instructor1].reduce((a, b) => a + b, 0);
+            const sum2 = res[name][instructor2].reduce((a, b) => a + b, 0);
+
+            return (sum2 / res[name][instructor2].length) - (sum1 / res[name][instructor1].length);
+          }
         }));
       })
-      .then(() => {
-        dispatch(endGenerating(schedules));
+      .finally(() => {
+        const schedules = genSchedules(filteredCourseList, values.gap, values.sortByGPA, getState().grades);
```

Comment:
instead of passing in `values.sortByGPA`, you should just sort out here.

Reply:
Fixed.